test/integration/yurttunnel_test.go

```diff
@@ -0,0 +1,208 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	anpserver "sigs.k8s.io/apiserver-network-proxy/pkg/server"
+
+	ta "github.com/alibaba/openyurt/pkg/yurttunnel/agent"
+	hw "github.com/alibaba/openyurt/pkg/yurttunnel/handlerwrapper"
+	tr "github.com/alibaba/openyurt/pkg/yurttunnel/handlerwrapper/tracerequest"
+	ts "github.com/alibaba/openyurt/pkg/yurttunnel/server"
+)
+
+const (
+	ResponseForRegularRequest = "Dummy Server"
+	DummyServerPort           = 9515
+	ServerMasterPort          = 9516
+	ServerMasterInsecurePort  = 9517
+	ServerAgentPort           = 9518
+	HTTPPath                  = "yurttunel-test"
+	RootCAFile                = "ca.pem"
+	GeneircCertFile           = "cert.pem"
+	GenericKeyFile            = "key.pem"
+	InterceptorServerUDSFile  = "interceptor-proxier.sock"
+)
+
+func genCAPool(t *testing.T, CAPath string) *x509.CertPool {
+	caCertPEM, err := ioutil.ReadFile(CAPath)
+	if err != nil {
+		t.Fatalf("fail to load the CA: %v", err)
+	}
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(caCertPEM)
+	if !ok {
+		t.Fatal("fail to append the ca PEM to cert pool")
+	}
+	return roots
+}
+
+func genCert(t *testing.T, certPath, keyPath string) tls.Certificate {
+	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		t.Fatalf("fail to load the dummy server cert: %v", err)
+	}
+	return cert
+}
+
+func startDummyServer(t *testing.T) {
+	m := http.NewServeMux()
+	m.HandleFunc("/"+HTTPPath, func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(ResponseForRegularRequest))
+	})
+
+	s := &http.Server{
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{
+				genCert(t, GeneircCertFile, GenericKeyFile),
+			},
+			ClientCAs: genCAPool(t, RootCAFile),
+		},
+		Addr:    fmt.Sprintf(":%d", DummyServerPort),
+		Handler: m,
+	}
+
+	klog.Info("[TEST] dummy-server is listening on :9515")
+	if err := s.ListenAndServeTLS("", ""); err != nil {
+		t.Fatalf("the dummy-server failed: %v", err)
+	}
+}
+
+func startDummyClient(t *testing.T, wg *sync.WaitGroup) {
+	defer wg.Done()
+	c := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				Certificates: []tls.Certificate{
+					genCert(t, GeneircCertFile, GenericKeyFile),
+				},
+				RootCAs: genCAPool(t, RootCAFile),
+			},
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("tcp", "127.0.0.1:9516")
+			},
+		},
+	}
+
+	r, err := http.NewRequest(http.MethodGet, "https://127.0.0.1:9515/"+HTTPPath, nil)
+	if err != nil {
+		t.Fatalf("fail to generate http request: %v", err)
+	}
+
+	rep, err := c.Do(r)
+	if err != nil {
+		t.Fatalf("fail to send request to the server: %v", err)
+	}
+	defer rep.Body.Close()
+
```

Comment:
how about add a snippet of http status code verification

Reply:
Yes. will switch to the status code